### ISSUE TITLE
Seal pooled events?

### ISSUE THREAD

**Issue by syranide** [facebook/react#5853]
Kind of related to the old #1612 and #5840, anyone trying to be smart and adding properties to a pooled non-persisted event will cause those to properties to remain in the pool, causing a memory leak of sorts. Could it make sense to `Object.seal(event)` the events, that way we ensure that users can't mess with the event object in bad or unsupported ways.

**Should probably do this instead:** https://github.com/facebook/react/issues/5853#issuecomment-171685664

**Comment by jimfb** [comment]
But you can't unseal when the user does persist the object, right?

**Comment by syranide** [comment]
@jimfb Yeah, that's obviously an issue. I would call it bad practice to modify events like that, so I personally wouldn't mind preventing it, but it might still be something we want to support. We could also clone instead of persisting events (same number of allocations), but it would obviously affect the API (`event = event.clone()`).

**Comment by jimfb** [comment]
Would it be sufficient to just detect the problem and warn?

We could check to verify that there are no extraneous properties on the event object after the event handler has completed, right?  Or we could wrap the event in a proxy object (in dev mode and where supported - firefox, edge, chrome-canary, etc) and warn if someone mutates an event before persisting it.

**Comment by syranide** [comment]
> Would it be sufficient to just detect the problem and warn? We could check to verify that there are no extraneous properties on the event object after the event handler has completed, right?

Yeah that should be fine too. Ultimately you're messing up the pooled events so it really should be an error. But then it couldn't be DEV-only and that's bad.

> Or we could wrap the event in a proxy object (in dev mode and where supported - firefox, edge, chrome-canary, etc) and warn if someone mutates an event before persisting it.

Tangentially, we could even use this to warn when accessing an event after it has been returned to the pool?

**Comment by jimfb** [comment]
> Yeah that should be fine too. Ultimately you're messing up the pooled events so it really should be an error. But then it couldn't be DEV-only and that's bad.

There are lots of things that I believe should be errors, but the javascript community seems to like things that fail soft.  That was one of my biggest culture shocks, coming from a java background where the community believes that things should always fail super hard.  I've given up on that fight, need to choose my battles :P.

 > Tangentially, we could even use this to warn when accessing an event after it has been returned to the pool?

We could do that anyway, with `Object.defineProperty`, right?  Since we know what properties exist on events.

**Comment by syranide** [comment]
> There are lots of things that I believe should be errors, but the javascript community seems to like things that fail soft. That was one of my biggest culture shocks, coming from a java background where the community believes that things should always fail super hard. I've given up on that fight, need to choose my battles :P.

I agree, but this is more because we don't want the perf hit in production I would say.

> We could do that anyway, with Object.defineProperty, right? Since we know what properties exist on events.

Yeah I guess so... but it's probably a little involved and it technically changes the runtime behavior (they are now getters/setters instead of values), I doubt anyone would notice but still. IMHO it's fine if DEV-warnings for coding practices only work in modern browsers.

PS. Hmm, didn't we have something like this? Perhaps we already do... I remember @spicyj working on something like it, but perhaps it never shipped.

**Comment by syranide** [comment]
@jimfb Good first bug?

**Comment by jimfb** [comment]
Yep.

**Comment by koba04** [comment]
I'll work on this after #5840.

**Comment by antoaravinth** [comment]
@jimfb Hi, I was trying to work on this issue. As you have mentioned:

>Would it be sufficient to just detect the problem and warn? We could check to verify that there are no extraneous properties on the event object after the event handler has completed, right?

I tried to figure out are there any "hooks" for finding when the event handler have been completed. While I navigated the React source I found that `ReactEventListener` has a method called `dispatchEvent` which is what firing the event and its associated call backs. However I couldn't able to make out, when the Synthetic events are getting created and also when the event handler execution is complete. 

Any help?
